Form for adding notes

dglebovskiy opened the pull request:
- Implemented a form for adding notes
- Added a component for adding notes
- Added new methods to notesService
- Added new states to state controller
- Updated the webapp layout

dglebovskiy committed:
Form for adding notes
- Implemented a form for adding notes
- Added a component for adding notes
- Added new methods to notesService
- Added new states to state controller
- Updated the webapp layout

evgeny-sav commented:
+1

alekseyC64 commented on notes_app/client/app/notes/notes_service.js:
```diff
@@ -4,9 +4,17 @@
+        return $http.get(api_path + id + '/');
+      },
+      'create': function(note) {
+        note['owner'] = '/api/v1/user/1/';
```
magic line. Not all users have id=1

dglebovskiy committed:
Removed magic

evgeny-sav commented:
+1

dglebovskiy committed:
Note sharing functionality
- Added dependencies for the selector component
- Added selector to note creation form
- Created user service to access user data in the selector

alekseyC64 commented on notes_app/client/notes.js:
```diff
@@ -3,25 +3,34 @@
+      'resolve': {
+        notes: function(notesService) {
+          return notesService.list().then(function(response) {
+            return response.data.objects;
```
it looks like service code - work with request.

alekseyC64 commented on notes_app/client/notes.js:
```diff
@@ -3,25 +3,34 @@
+          return notesService.list().then(function(response) {
+            return response.data.objects;
+          }).catch(function(response) {
+            console.error('Error when loading notes');
```
please move `.catch` to the service layer

alekseyC64 commented on notes_app/notes/api.py:
```diff
@@ -112,3 +112,7 @@ class Meta:
         authentication = BasicAuthentication()
+
+    def hydrate(self, bundle):
+        bundle.obj.owner = bundle.request.user
```
assuming this code, does test of notes still work?

dglebovskiy committed:
Moved request handling to services

alekseyC64 commented on notes_app/client/app/notes/notes_service.js:
```diff
@@ -1,12 +1,40 @@
-      'getList': function() {
-        return $http.get('http://localhost:8000/api/v1/note/');
+      'list': function() {
+        return new Promise(function(fulfill, reject) {
```
Angular has it's own implementation of promises - https://docs.angularjs.org/api/ng/service/$q

alekseyC64 commented on notes_app/client/app/users/user_service.js:
```diff
@@ -0,0 +1,22 @@
+    var api_path = 'http://localhost:8000/api/v1/user/';
+    return {
+      'list': function() {
+        return new Promise(function(fulfill, reject) {
```
the same as above

alekseyC64 commented on notes_app/client/app/notes/note_add_form.component.js:
```diff
@@ -0,0 +1,25 @@
+    this.users = [];
+
+    userService.list().then(function(response) {
+      self.users = response.objects;
```
I think it should be re-implemented considering last service changes

dglebovskiy committed:
Form updates
- Added a component for alert messages
- Form now locks while awaiting the server response
- Form displays an alert with the request state
- Notes service now uses native angular promises (returned by $http)
- Notes service now returns $http promises, results will be handled in controllers

dglebovskiy committed:
Service logic cleanup

alekseyC64 commented on notes_app/client/app/notes/notes_service.js:
```diff
@@ -1,12 +1,38 @@
+          return response.data;
+        }).catch(function errorHandler(response) {
+          $log.error('Problem with getting the user list from server')
+          return {};
```
`list` shouldn't return an object, I think. Maybe empty array?

alekseyC64 commented on notes_app/client/app/notes/notes_service.js:
```diff
@@ -1,19 +1,38 @@
+          }
+        ).catch(function errorHandler(response) {
+          $log.error('Problem with getting the user from server');
+          return {};
```
I think here empty object is confusing. Code will wait an object, and if there is no object - it's better to return null, what do you think?

alekseyC64 commented on notes_app/client/app/notes/notes_service.js:
```diff
@@ -1,19 +1,38 @@
       'read': function(id) {
-        return $http.get(api_path + id + '/');
+        return $http.get(api_path + id + '/').then(
+          function successHandler(response) {
```
function name doesn't required

dglebovskiy committed:
More service fixes:
- fixed misleading error messages in notes service
- list api methods now return empty list on errors
- item api methods now return null on errors
- error alert now is correctly displayed on failure
